### PR TITLE
show squiggly line when import class not in build target

### PR DESCRIPTION
**Pull request by idankWix**

### CODE REVIEW

File: base/src/com/google/idea/blaze/base/sync/BlazeSyncTask.java [hunk 1]
```diff
@@ -114,11 +116,7 @@
 import com.intellij.openapi.util.io.FileUtil;
 import com.intellij.openapi.vfs.VirtualFileManager;
 import java.io.File;
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
```
Comment (or-shachar) [inline base/src/com/google/idea/blaze/base/sync/BlazeSyncTask.java]: @idankWix pretty sure Google convention is not to have wildcard imports...

File: base/src/com/google/idea/blaze/base/ui/problems/BlazeProblemsView.java [hunk 2]
```diff
@@ -34,25 +34,24 @@
...
 import java.io.File;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.EnumSet;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
-import java.util.StringTokenizer;
-import java.util.UUID;
+import java.util.*;
```
Comment (or-shachar) [inline base/src/com/google/idea/blaze/base/ui/problems/BlazeProblemsView.java]: same

**Comment by idankWix** [comment]
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here (e.g. `I signed it!`) and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google.com/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).

I signed it!

File: base/src/com/google/idea/blaze/base/run/BlazeBeforeRunCommandHelper.java [hunk 3]
```diff
@@ -65,7 +67,7 @@ private BlazeBeforeRunCommandHelper() {}
         (BlazeCommandRunConfigurationCommonState) configuration.getHandler().getState();
     WorkspaceRoot workspaceRoot = WorkspaceRoot.fromProject(project);
     ProjectViewSet projectViewSet = ProjectViewManager.getInstance(project).getProjectViewSet();
-
+      ServiceManager.getService(ImportProblemContainerService.class).resetIssues();
```
Comment (or-shachar) [inline base/src/com/google/idea/blaze/base/run/BlazeBeforeRunCommandHelper.java]: @hmemcpy - is there an event system in Intellij such that instead of triggering an action here, we can broadcast an event that will cause other processes to be triggered?

**Comment by idankWix** [comment]
before
after

**Comment by ittaiz** [comment]
@brendandouglas wdyt?

File: base/src/com/google/idea/blaze/base/lang/buildfile/actions/BuildFileModifierImpl.java [hunk 4]
```diff
@@ -53,7 +67,110 @@ public boolean addRule(Project project, Label newRule, Kind ruleKind) {
     return true;
   }
 
-  private PsiElement createRule(Project project, Kind ruleKind, String ruleName) {
+  @Override
+  public boolean addDepToRule(Project project,
```
Comment (or-shachar) [inline base/src/com/google/idea/blaze/base/lang/buildfile/actions/BuildFileModifierImpl.java]: This can greatly be simplified if #477 was solved.

File: base/src/com/google/idea/blaze/base/lang/buildfile/actions/BuildFileModifierImpl.java [hunk 5]
```diff
@@ -53,7 +67,110 @@ public boolean addRule(Project project, Label newRule, Kind ruleKind) {
...
+    private java.util.Optional<Argument.Keyword> findTargetDepsAttribute(PsiElement psiElement) {
+        List<Argument.Keyword> allChildKeywords = PsiUtils.findAllChildrenOfClassRecursive(psiElement, Argument.Keyword.class);
+        Stream<Argument.Keyword> depsKeywords = allChildKeywords.stream().filter(keyword -> keyword.getName().equals("deps"));
+        return depsKeywords.findFirst();
+    }
+
+    private void addTargetToDepsIfNeeded(Argument.Keyword depsKeyword, Label targetToBeAdded, Project project) {
+        PsiElement depsElement = depsKeyword.getPsiChild(BuildElementTypes.LIST_LITERAL, null);
+        if(depsDoesntHaveWhiteSpace(depsElement)){
+            addWhiteSpaceElement(project, depsElement, "\n    ");
```
Comment (or-shachar) [inline base/src/com/google/idea/blaze/base/lang/buildfile/actions/BuildFileModifierImpl.java]: Can we use https://github.com/bazelbuild/intellij/blob/master/base/src/com/google/idea/blaze/base/buildmodifier/BuildFileFormatter.java for final formatting?

**Comment by idankWix** [comment]
@brendandouglas can you pls go over the PR ?

File: base/src/com/google/idea/blaze/base/lang/buildfile/actions/BuildFileModifierImpl.java (hunk 5 above)
Comment (idankWix) [reply base/src/com/google/idea/blaze/base/lang/buildfile/actions/BuildFileModifierImpl.java]: todo

File: base/BUILD [hunk 6]
```diff
@@ -19,6 +19,7 @@ java_library(
         "//sdkcompat",
         "//third_party/auto_value",
         "@error_prone_annotations//jar",
+        "//third_party/scala",
```
Comment (brendandouglas) [inline base/BUILD]: The 'base' module is used across all IDEs, and can't have a language specific dep (unless that language is supported in all JetBrains IDEs).

**Comment by brendandouglas** [comment]
Thanks for the pull request. I haven't had a chance to look over it yet, but will try to get to it within the next day or two.

File: base/src/com/google/idea/blaze/base/ui/problems/ImportIssueResolver.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import com.intellij.psi.PsiManager;
+
+import java.util.regex.Pattern;
+
+public class ImportIssueResolver {
+
+    private static final String NEW_LINE = "\n";
+    private static String missingImportRegEx = "object .* is not a member of package .*";
+    private static String wildCardOrStaticRegEx = "package .* does not exist";
+    private static String javaStaticImport = "static import only from classes and interfaces";
```
Comment (brendandouglas) [inline base/src/com/google/idea/blaze/base/ui/problems/ImportIssueResolver.java]: Shouldn't problems like this already be annotated in the IDE?

**Comment by brendandouglas** [comment]
Thanks again for the PR. I think the general approach of annotating missing imports/deps and providing a quick-fix action is good. Ideally, we wouldn't need to run a bazel build to find these.

We have a separate system for this internally. Bazel outputs a error with a suggested fix such as "Command to add missing dependencies: '[command line script]'". We add a hyperlink to the bazel console view (also linked in the bazel problems view) to run the script and apply the fix.

Otherwise, I'm not sure when we'd be able to work on this PR, since there's a fair bit of work involved applying it upstream. Some examples: I'd prefer to use buildozer directly rather than manually modifying BUILD files, the language-specific bits need to be moved to the appropriate modules, we'd need to add some tests and documentation, and make some style changes.

File: base/src/com/google/idea/blaze/base/ui/problems/ImportIssueResolver.java (hunk 7 above)
Comment (idankWix) [reply base/src/com/google/idea/blaze/base/ui/problems/ImportIssueResolver.java]: they are not currently

**Comment by or-shachar** [comment]
@brendandouglas - how do we proceed on this? 

>  Ideally, we wouldn't need to run a bazel build to find these.
What are the alternatives?

> We have a separate system for this internally.
Great! but we need to enrich this public plugin as well.. And this is our contribution to make it better and complete. Can you elaborate on the tool? is it something Wix can use?

> Otherwise, I'm not sure when we'd be able to work on this PR, since there's a fair bit of work involved applying it upstream. 
I'm a bit confused by this. What does it mean?

>  I'd prefer to use buildozer directly rather than manually modifying BUILD files, the language-specific bits need to be moved to the appropriate modules
See my suggestions here: #477  - would love to hear your comment . If you need our contribution here - we can do it. 

>  we'd need to add some tests and documentation, and make some style changes.
What is the style guide? Where can we add documentation?


===
Bottom line - From the belief in bazel as a build tool and Intellij as our IDE, and from an Open Source approach - we are willing to work hard and contribute but we need your full support on it, but we need your hand in this.

Would really like to know what are the concrete steps you want to do in order to push this plugin forward.

Thanks!
cc: @ittaiz

**Comment by brendandouglas** [comment]
@or-shachar, our preferred way of implementing this kind of feature would be via 'add_deps' support at bazel build time (there's some discussion of this in bazelbuild/bazel/issues/4846 and bazelbuild/bazel/issues/4584).

That moves responsibility for determining the appropriate fix to the build, rather than having a separate layer in the IDE. If that was implemented in bazel, we'd simply parse the relevant command in bazel's stdout/stderr, and add an entry to the 'bazel problems' view, as well as adding a link to the console output.

> > Otherwise, I'm not sure when we'd be able to work on this PR, since there's a fair bit of work involved
> > applying it upstream.
> 
> I'm a bit confused by this. What does it mean?

This github project is just a mirror of our internal project. Applying patches involves:
- rewriting the PR to conform to our coding practices, style guide, etc.
- adding appropriate unit and integration tests and safeguards to lessen the chance of internal regressions, especially related to performance.

All of this takes time, so we prefer to iron out the details of a potential contribution beforehand, to save effort (see the 'contributions' section of the top level readme).

File: base/BUILD (hunk 6 above)
Comment (idankWix) [reply base/BUILD]: fixed by moving to async messaging
